allennlp_models/vision/dataset_readers/vgqa.py

```diff
@@ -0,0 +1,258 @@
+
+        question_dicts = list(self.shard_iterable(questions))
+        processed_images: Iterable[Optional[Tuple[Tensor, Tensor]]]
+        if self.produce_featurized_images:
+            # It would be much easier to just process one image at a time, but it's faster to process
+            # them in batches. So this code gathers up instances until it has enough to fill up a batch
+            # that needs processing, and then processes them all.
+
+            filenames = [f"{question_dict['image_id']}.jpg" for question_dict in question_dicts]
+            logger.info("images size: %s", len(self.images))
+            try:
+                processed_images = self._process_image_paths(
+                    self.images[filename] for filename in filenames
+                )
+            except KeyError as e:
+                print(self.images)
+                missing_id = e.args[0]
+                raise KeyError(
+                    missing_id,
+                    f"We could not find an image with the id {missing_id}. "
+                    "Because of the size of the image datasets, we don't download them automatically. "
+                    "Please go to https://visualqa.org/download.html, download the datasets you need, "
+                    "and set the image_dir parameter to point to your download location. This dataset "
+                    "reader does not care about the exact directory structure. It finds the images "
+                    "wherever they are.",
+                )
+        else:
+            processed_images = [None for _ in range(len(question_dicts))]
+
+        logger.info("Reading the dataset")
+        failed_instances_count = 0
+        attempted_instances_count = 0
+        for qa, processed_image in zip(question_dicts, processed_images):
+            question = qa["question"]
+            answer = utils.preprocess_answer(qa["answer"])
+            qa_id = qa["qa_id"]
+
+            instance = self.text_to_instance(
+                qa_id,
+                question,
+                answer,
+                processed_image,
+            )
+
+            attempted_instances_count += 1
+            if instance is not None:
+                yield instance
+            else:
+                failed_instances_count += 1
+
+        failed_instances_fraction = failed_instances_count / attempted_instances_count
+        logger.warning(f"{failed_instances_fraction*100:.0f}% of instances failed.")
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        qa_id: int,
+        question: str,
+        answer: str,
+        image: Union[str, Tuple[Tensor, Tensor]],
+        use_cache: bool = True,
+        keep_impossible_questions: bool = True,
+    ) -> Optional[Instance]:
+        question_field = TextField(self._tokenizer.tokenize(question), None)
+
+        fields: Dict[str, Field] = {
+            "question": question_field,
+        }
+
+        if image is None:
+            return None
```

Comment:
Is `image` ever `None` without it being an error? The type annotation doesn't even allow that.

Reply:
That's fair, I was following what VQA was doing, but it doesn't make sense in this context.